File: user-service/src/test/java/rs/raf/user_service/AuthControllerTest.java
package rs.raf.user_service;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import rs.raf.user_service.controller.AuthController;
import rs.raf.user_service.controller.UserController;
import rs.raf.user_service.dto.ActivationRequestDto;
import rs.raf.user_service.dto.LoginRequestDto;
import rs.raf.user_service.dto.LoginResponseDto;
import rs.raf.user_service.dto.RequestPasswordResetDto;
import rs.raf.user_service.service.AuthService;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.*;

public class AuthControllerTest {

    private AuthService authService;
    private AuthController authController;
    private UserController userController;

    @BeforeEach
    public void setup() {
        authService = mock(AuthService.class);
        authController = new AuthController(authService);
        userController = new UserController();
    }

    @Test
    public void testClientLogin_Success() {
        LoginRequestDto request = new LoginRequestDto();
        request.setEmail("test@example.com");
        request.setPassword("password");

        String expectedToken = "valid_jwt_token";
        when(authService.authenticateClient(request.getEmail(), request.getPassword())).thenReturn(expectedToken);

        ResponseEntity<?> responseEntity = authController.clientLogin(request);

        assertEquals(HttpStatus.OK, responseEntity.getStatusCode());
        assertNotNull(responseEntity.getBody());
        assertInstanceOf(LoginResponseDto.class, responseEntity.getBody());

        LoginResponseDto responseDTO = (LoginResponseDto) responseEntity.getBody();
        assertEquals(expectedToken, responseDTO.getToken());
    }

    @Test
    public void testClientLogin_InvalidCredentials() {
        LoginRequestDto request = new LoginRequestDto();
        request.setEmail("invalid@example.com");
        request.setPassword("wrongpassword");

        when(authService.authenticateClient(request.getEmail(), request.getPassword())).thenReturn(null);

        ResponseEntity<?> responseEntity = authController.clientLogin(request);

        assertEquals(HttpStatus.UNAUTHORIZED, responseEntity.getStatusCode());
        assertEquals("Bad credentials", responseEntity.getBody());
    }

    @Test
    public void testEmployeeLogin_Success() {
        LoginRequestDto request = new LoginRequestDto();
        request.setEmail("test@example.com");
        request.setPassword("password");

        String expectedToken = "valid_jwt_token";
        when(authService.authenticateEmployee(request.getEmail(), request.getPassword())).thenReturn(expectedToken);

        ResponseEntity<?> responseEntity = authController.employeeLogin(request);

        assertEquals(HttpStatus.OK, responseEntity.getStatusCode());
        assertNotNull(responseEntity.getBody());
        assertInstanceOf(LoginResponseDto.class, responseEntity.getBody());

        LoginResponseDto responseDTO = (LoginResponseDto) responseEntity.getBody();
        assertEquals(expectedToken, responseDTO.getToken());
    }

    @Test
    public void testEmployeeLogin_InvalidCredentials() {
        LoginRequestDto request = new LoginRequestDto();
        request.setEmail("invalid@example.com");
        request.setPassword("wrongpassword");

        when(authService.authenticateEmployee(request.getEmail(), request.getPassword())).thenReturn(null);

        ResponseEntity<?> responseEntity = authController.employeeLogin(request);

        assertEquals(HttpStatus.UNAUTHORIZED, responseEntity.getStatusCode());
        assertEquals("Bad credentials", responseEntity.getBody());
    }

    @Test
    void testRequestPasswordReset_Success() {
        RequestPasswordResetDto requestPasswordResetDTO = new RequestPasswordResetDto();
        requestPasswordResetDTO.setEmail("test@example.com");

        ResponseEntity<Void> response = authController.requestPasswordReset(requestPasswordResetDTO);

        assertNotNull(response);
        assertEquals(HttpStatus.OK, response.getStatusCode());
    }

    @Test
    void testRequestPasswordReset_Failure() {
        RequestPasswordResetDto requestPasswordResetDTO = new RequestPasswordResetDto();
        requestPasswordResetDTO.setEmail("test@example.com");

        doThrow(new RuntimeException()).when(authService).requestPasswordReset(anyString());

        ResponseEntity<Void> response = authController.requestPasswordReset(requestPasswordResetDTO);

        assertNotNull(response);
        assertEquals(HttpStatus.NOT_FOUND, response.getStatusCode());
    }

    @Test
    void testResetPassword_Success() {
        ActivationRequestDto dto = new ActivationRequestDto();
        dto.setToken("valid_token");
        dto.setPassword("newPassword");

        ResponseEntity<Void> response = authController.activateUser(dto);

        assertNotNull(response);
        assertEquals(HttpStatus.OK, response.getStatusCode());
    }

    @Test
    public void testActivateUser_Success() {
        ActivationRequestDto request = new ActivationRequestDto();
        request.setToken("valid-token");
        request.setPassword("newPassword123");

        ResponseEntity<Void> response = authController.activateUser(request);

        assertEquals(HttpStatus.OK, response.getStatusCode());

        // Provera da li je servis pozvan
        verify(authService, times(1)).setPassword("valid-token", "newPassword123");
    }

    @Test
    public void testActivateUser_InvalidToken() {
        ActivationRequestDto request = new ActivationRequestDto();
        request.setToken("invalid-token");
        request.setPassword("newPassword123");

        doThrow(new RuntimeException("Invalid token.")).when(authService).setPassword("invalid-token", "newPassword123");

        ResponseEntity<Void> response = authController.activateUser(request);

        assertEquals(HttpStatus.NOT_FOUND, response.getStatusCode());

        verify(authService, times(1)).setPassword("invalid-token", "newPassword123");
    }

    @Test
    public void testActivateUser_ExpiredToken() {
        ActivationRequestDto request = new ActivationRequestDto();
        request.setToken("expired-token");
        request.setPassword("newPassword123");

        doThrow(new RuntimeException("Expired token.")).when(authService).setPassword("expired-token", "newPassword123");

        ResponseEntity<Void> response = authController.activateUser(request);

        assertEquals(HttpStatus.NOT_FOUND, response.getStatusCode());

        verify(authService, times(1)).setPassword("expired-token", "newPassword123");
    }
}

File: user-service-kotlin/src/test/kotlin/pack/userservicekotlin/controller/AuthControllerTest.kt
package pack.userservicekotlin.controller

import arrow.core.Either
import com.fasterxml.jackson.databind.ObjectMapper
import org.junit.jupiter.api.Test
import org.mockito.Mockito.`when`
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest
import org.springframework.http.MediaType
import org.springframework.test.context.bean.override.mockito.MockitoBean
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.post
import pack.userservicekotlin.arrow.AuthServiceError
import pack.userservicekotlin.domain.dto.activity_code.ActivationRequestDto
import pack.userservicekotlin.domain.dto.activity_code.RequestPasswordResetDto
import pack.userservicekotlin.domain.dto.external.CheckTokenDto
import pack.userservicekotlin.domain.dto.login.LoginRequestDto
import pack.userservicekotlin.service.AuthService
import pack.userservicekotlin.utils.JwtTokenUtil

@WebMvcTest(AuthController::class)
@AutoConfigureMockMvc(addFilters = false)
class AuthControllerTest {
    @Autowired
    lateinit var mockMvc: MockMvc

    @Autowired
    lateinit var objectMapper: ObjectMapper

    @MockitoBean
    lateinit var authService: AuthService

    @MockitoBean
    lateinit var jwtTokenUtil: JwtTokenUtil

    @Test
    fun `clientLogin should return token when credentials are valid`() {
        val loginRequest = LoginRequestDto(email = "client@example.com", password = "password123")
        val token = "mocked-jwt-token"

        `when`(authService.authenticateClient(loginRequest.email, loginRequest.password))
            .thenReturn(Either.Right(token))

        mockMvc
            .post("/api/auth/login/client") {
                contentType = MediaType.APPLICATION_JSON
                content = objectMapper.writeValueAsString(loginRequest)
            }.andExpect {
                status { isOk() }
                jsonPath("$.token") { value(token) }
            }
    }

    @Test
    fun `clientLogin should return 401 when credentials are invalid`() {
        val loginRequest = LoginRequestDto(email = "client@example.com", password = "wrongpassword")

        `when`(authService.authenticateClient(loginRequest.email, loginRequest.password))
            .thenReturn(Either.Left(AuthServiceError.InvalidCredentials))

        mockMvc
            .post("/api/auth/login/client") {
                contentType = MediaType.APPLICATION_JSON
                content = objectMapper.writeValueAsString(loginRequest)
            }.andExpect {
                status { isUnauthorized() }
                content { string("Bad credentials") }
            }
    }

    @Test
    fun `employeeLogin should return token when credentials are valid`() {
        val loginRequest = LoginRequestDto(email = "employee@example.com", password = "password123")
        val token = "mocked-jwt-token"

        `when`(authService.authenticateEmployee(loginRequest.email, loginRequest.password))
            .thenReturn(Either.Right(token))

        mockMvc
            .post("/api/auth/login/employee") {
                contentType = MediaType.APPLICATION_JSON
                content = objectMapper.writeValueAsString(loginRequest)
            }.andExpect {
                status { isOk() }
                jsonPath("$.token") { value(token) }
            }
    }

    @Test
    fun `employeeLogin should return 401 when credentials are invalid`() {
        val loginRequest = LoginRequestDto(email = "employee@example.com", password = "wrongpassword")

        `when`(authService.authenticateEmployee(loginRequest.email, loginRequest.password))
            .thenReturn(Either.Left(AuthServiceError.InvalidCredentials))

        mockMvc
            .post("/api/auth/login/employee") {
                contentType = MediaType.APPLICATION_JSON
                content = objectMapper.writeValueAsString(loginRequest)
            }.andExpect {
                status { isUnauthorized() }
                content { string("Bad credentials") }
            }
    }

    /** Calling external services */

//    @Test
//    fun `requestPasswordReset should return 200 when email exists`() {
//        val request = RequestPasswordResetDto(email = "user@example.com")
//
//        `when`(authService.requestPasswordReset(request.email))
//            .thenReturn(Either.Right(Unit))
//
//        mockMvc
//            .post("/api/auth/request-password-reset") {
//                contentType = MediaType.APPLICATION_JSON
//                content = objectMapper.writeValueAsString(request)
//            }.andExpect {
//                status { isOk() }
//            }
//    }

    @Test
    fun `requestPasswordReset should return 404 when email does not exist`() {
        val request = RequestPasswordResetDto(email = "nonexistent@example.com")

        `when`(authService.requestPasswordReset(request.email))
            .thenReturn(Either.Left(AuthServiceError.UserNotFound))

        mockMvc
            .post("/api/auth/request-password-reset") {
                contentType = MediaType.APPLICATION_JSON
                content = objectMapper.writeValueAsString(request)
            }.andExpect {
                status { isNotFound() }
            }
    }

    /** Calling external services */

//    @Test
//    fun `checkToken should return 200 when token is valid`() {
//        val request = CheckTokenDto(token = "valid-token")
//
//        `when`(authService.checkToken(request.token))
//            .thenReturn(Either.Right(Unit))
//
//        mockMvc
//            .post("/api/auth/check-token") {
//                contentType = MediaType.APPLICATION_JSON
//                content = objectMapper.writeValueAsString(request)
//            }.andExpect {
//                status { isOk() }
//            }
//    }

    @Test
    fun `checkToken should return 404 when token is invalid`() {
        val request = CheckTokenDto(token = "invalid-token")

        `when`(authService.checkToken(request.token))
            .thenReturn(Either.Left(AuthServiceError.InvalidToken))

        mockMvc
            .post("/api/auth/check-token") {
                contentType = MediaType.APPLICATION_JSON
                content = objectMapper.writeValueAsString(request)
            }.andExpect {
                status { isNotFound() }
            }
    }

    /** Calling external services */

//    @Test
//    fun `activateUser should return 200 when token and password are valid`() {
//        val request = ActivationRequestDto(token = "valid-token", password = "newpassword123")
//
//        `when`(authService.setPassword(request.token, request.password))
//            .thenReturn(Either.Right(Unit))
//
//        mockMvc
//            .post("/api/auth/set-password") {
//                contentType = MediaType.APPLICATION_JSON
//                content = objectMapper.writeValueAsString(request)
//            }.andExpect {
//                status { isOk() }
//            }
//    }

    @Test
    fun `activateUser should return 404 when token is invalid`() {
        val request = ActivationRequestDto(token = "invalid-token", password = "newpassword123")

        `when`(authService.setPassword(request.token, request.password))
            .thenReturn(Either.Left(AuthServiceError.InvalidToken))

        mockMvc
            .post("/api/auth/set-password") {
                contentType = MediaType.APPLICATION_JSON
                content = objectMapper.writeValueAsString(request)
            }.andExpect {
                status { isNotFound() }
            }
    }
}
